tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb

```diff
@@ -0,0 +1,732 @@
+    "        }\n",
+    "\n",
+    "    def __init__(self, sample_rate):\n",
+    "        super().__init__()\n",
+    "        self._sample_rate = sample_rate\n",
+    "        self.output = True\n",
+    "        \n",
+    "    def __iter__(self):\n",
+    "        return self\n",
+    "    \n",
+    "    def __next__(self):\n",
+    "        if not self.output:\n",
+    "            raise StopIteration\n",
+    "        self.output = False\n",
+    "        return torch.as_tensor(self.signal, dtype=torch.float32), \\\n",
+    "               torch.as_tensor(self.signal_shape, dtype=torch.int64)\n",
+    "        \n",
+    "    def set_signal(self, signal):\n",
+    "        self.signal = signal.astype(np.float32)/32768.\n",
+    "        self.signal_shape = self.signal.size\n",
+    "        self.output = True\n",
+    "\n",
+    "    def __len__(self):\n",
+    "        return 1"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data_layer = AudioDataLayer(sample_rate=mbn_cfg.train_ds.sample_rate)\n",
+    "data_loader = DataLoader(data_layer, batch_size=1, collate_fn=data_layer.collate_fn)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# inference method for audio signal (single instance)\n",
+    "def infer_signal(model, signal):\n",
+    "    data_layer.set_signal(signal)\n",
+    "    batch = next(iter(data_loader))\n",
+    "    audio_signal, audio_signal_len = batch\n",
+    "    audio_signal, audio_signal_len = audio_signal.to(model.device), audio_signal_len.to(model.device)\n",
+    "    logits = model.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)\n",
+    "    return logits"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "we don't include postprocessing techniques here. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# class for streaming frame-based ASR\n",
+    "# 1) use reset() method to reset FrameASR's state\n",
+    "# 2) call transcribe(frame) to do ASR on\n",
+    "#    contiguous signal's frames\n",
+    "class FrameASR:\n",
+    "    \n",
+    "    def __init__(self, model_definition,\n",
+    "                 frame_len=2, frame_overlap=2.5, \n",
+    "                 offset=0):\n",
+    "        '''\n",
+    "        Args:\n",
+    "          frame_len: frame's duration, seconds\n",
+    "          frame_overlap: duration of overlaps before and after current frame, seconds\n",
+    "          offset: number of symbols to drop for smooth streaming\n",
+    "        '''\n",
+    "        self.task = model_definition['task']\n",
+    "        self.vocab = list(model_definition['labels'])\n",
+    "        self.vocab.append(\"_\")\n",
```

Comment:
Add a comment why this is being added - MBN doesnt have blanks right?

Reply:
@titu1994  Right MBN doesn't have blanks AFIK. I got it from your draft and thought it could detect UNK words, but it actually cannot. When I say words outside the vocab, it gives me "closest" prediction not "_". 
What's your opinion?  I'm using 35 classes maybe I should switch 10+2?

Reply:
Right I used the _ token as a way to prevent class output in 10+2 case. I think it can be removed safely in VAD + MBN case (but give it a try).

Reply:
But VAD+MBN cannot handle words outside the vocab, right?

Reply:
Neither of these models can handle out if vocab, so let's note it and keep this with a comment that it's for oov

Reply:
okay then let's keep 35. wait...it's for oov and the models cannot handle oov. Therefore, we expect it won't give "_" so it's useless. Do I understand it right?